pkg/controller/apiextensions/composite/api.go

```diff
@@ -125,6 +155,68 @@ func (r *APISelectorResolver) ResolveSelector(ctx context.Context, cp resource.C
 	return errors.New(errNoCompatibleComposition)
 }
 
+// NewAPIDefaultCompositionSelector returns a APIDefaultCompositionSelector.
+func NewAPIDefaultCompositionSelector(c client.Client, ref v1.ObjectReference, r event.Recorder) *APIDefaultCompositionSelector {
+	return &APIDefaultCompositionSelector{client: c, defRef: ref, recorder: r}
+}
+
+// APIDefaultCompositionSelector selects the default composition referenced in
+// the definition of the resource if neither a reference nor selector is given
+// in composite resource.
+type APIDefaultCompositionSelector struct {
+	client   client.Client
+	defRef   v1.ObjectReference
+	recorder event.Recorder
+}
+
+// SelectComposition selects the default compositionif neither a reference nor
+// selector is given in composite resource.
+func (s *APIDefaultCompositionSelector) SelectComposition(ctx context.Context, cp resource.Composite) error {
+	if cp.GetCompositionReference() != nil || cp.GetCompositionSelector() != nil {
+		return nil
+	}
+	def := &v1alpha1.InfrastructureDefinition{}
+	if err := s.client.Get(ctx, meta.NamespacedNameOf(&s.defRef), def); err != nil {
+		return errors.Wrap(err, errGetInfraDef)
+	}
+	if def.Spec.DefaultCompositionRef == nil {
+		return nil
+	}
+	cp.SetCompositionReference(def.Spec.DefaultCompositionRef)
+	s.recorder.Event(cp, event.Normal(reasonCompositionSelection, "Default composition has been selected"))
+	return nil
+}
+
+// NewEnforcedCompositionSelector returns a EnforcedCompositionSelector.
+func NewEnforcedCompositionSelector(def v1alpha1.InfrastructureDefinition, r event.Recorder) *EnforcedCompositionSelector {
+	return &EnforcedCompositionSelector{def: def, recorder: r}
+}
+
+// EnforcedCompositionSelector , if it's given, selects the enforced composition
+// on the definition for all composite instances.
+type EnforcedCompositionSelector struct {
+	def      v1alpha1.InfrastructureDefinition
+	recorder event.Recorder
+}
+
+// SelectComposition selects the enforced composition if it's given in definition.
+func (s *EnforcedCompositionSelector) SelectComposition(_ context.Context, cp resource.Composite) error {
+	// We don't need to fetch the InfrastructureDefinition at every reconcile
+	// because enforced composition ref is immutable as opposed to default
+	// composition ref.
+	if s.def.Spec.EnforcedCompositionRef == nil {
+		return nil
+	}
+	// If the composition is already chosen, we don't need to check for compatibility
+	// as its target type reference is immutable.
+	if cp.GetCompositionReference().String() == s.def.Spec.EnforcedCompositionRef.String() {
+		return nil
+	}
```

Comment:
Do we need this branch? It seems like if we removed it we'd have a no-op `SetCompositionReference` below, so the only difference is we might emit an event despite the enforced composition already being set.

Reply:
Yeah, I believe the event should be fired only when it's actually fired to keep the chronological order of events digestable. I'd like to avoid a situation where there are events on the object but one of every two event is `enforced composition is selected` whereas it's actually selected only if there is a drift.